src/esp32/pwm.rs

```diff
@@ -209,123 +120,241 @@ impl<'a> PwmDriver<'a> {
-        Ok(())
+    fn next_available_timer(&mut self, frequency_hz: u32) -> Result<usize, Esp32PwmError> {
+        if let Some(timer) = self.find_timer_by_frequency(frequency_hz) {
+            return Ok(timer);
+        }
+        let timer_number = self
+            .timer_allocation
+            .iter()
+            .enumerate()
+            .find_map(|(i, t)| if t.count == 0 { Some(i) } else { None })
+            .ok_or(Esp32PwmError::NoTimersAvailable)?;
+        self.timer_allocation[timer_number].set_frequency(frequency_hz)?;
+        Ok(timer_number)
     }
 
-    fn number_of_channels_by_timer(&self, timer_number: usize) -> usize {
-        self.timer_channels[timer_number].len()
+    fn find_timer_by_frequency(&mut self, frequency_hz: u32) -> Option<usize> {
+        self.timer_allocation
+            .iter()
+            .position(|t| t.frequency == frequency_hz)
     }
 
-    fn take_timer(
+    fn set_timer_frequency(
         &mut self,
+        timer_number: usize,
         frequency_hz: u32,
         channel: PwmChannel,
     ) -> Result<usize, Esp32PwmError> {
-        match self.find_timer_by_frequency(frequency_hz) {
-            Some(timer_number) => {
-                self.timer_channels[timer_number].push(channel);
-                Ok(timer_number)
-            }
-            None => {
-                if self.timer_frequencies.iter().all(|&x| x.is_some()) {
-                    Err(Esp32PwmError::NoTimersAvailable(
-                        self.timer_frequencies[0].unwrap(),
-                        self.timer_frequencies[1].unwrap(),
-                        self.timer_frequencies[2].unwrap(),
-                        self.timer_frequencies[3].unwrap(),
-                    ))
-                } else {
-                    let timer_number = match self.timer_frequencies.iter().position(|x| x.is_none())
-                    {
-                        Some(timer_number) => timer_number,
-                        None => unreachable!(),
-                    };
-                    self.timer_frequencies[timer_number] = Some(frequency_hz);
-                    self.timer_channels[timer_number].push(channel);
-                    Ok(timer_number)
-                }
-            }
+        if timer_number > 3 {
+            return Err(Esp32PwmError::InvalidTimerNumber(timer_number as i32));
         }
-    }
 
-    fn drop_timer(&mut self, timer_number: usize, channel: PwmChannel) {
-        let timer_channels = &mut self.timer_channels[timer_number];
-        if let Some(index) = timer_channels.iter().position(|&x| x == channel) {
-            timer_channels.swap_remove(index);
-            if self.timer_channels[timer_number].is_empty() {
-                self.timer_frequencies[timer_number] = None;
-            };
-        };
-    }
+        self.timer_allocation[timer_number].dec();
+        let res = self.timer_allocation[timer_number].set_frequency(frequency_hz);
+        self.timer_allocation[timer_number].inc();
```

Comment:
Should this `inc` happen even if this function ultimately returns a `NoTimersAvailable` error, down on 299?

Reply:
Also, isn't it already being incremented within `bind_channel_to_timer` at Line 329?

Reply:
The logic here is to temporally decrement the timer count so we can find out if other channels are bind to it. 
So we have 3 Options:
1) One channel is bind so we update the frequency and the count remain unchanged
2) Another Timer is allocated and we decrement the old timer and increment the new one
3) No other timer can be allocated the channel remain bound to the previous timer and nothing happens

Reply:
Thanks, with those cases listed out I see how it works. A comment summarizing what you've said here might be worthwhile for future us.